sys/net/gnrc/application_layer/coap/gcoap.c

```diff
@@ -0,0 +1,363 @@
+
+    while (1) {
+        msg_receive(&msg_rcvd);
+
+        switch (msg_rcvd.type) {
+            case GNRC_NETAPI_MSG_TYPE_RCV:
+                /* find client from UDP destination port */
+                DEBUG("coap: GNRC_NETAPI_MSG_TYPE_RCV\n");
+                pkt = (gnrc_pktsnip_t *)msg_rcvd.content.ptr;
+                if (pkt->type != GNRC_NETTYPE_UNDEF) {
+                    gnrc_pktbuf_release(pkt);
+                    break;
+                }
+                udp_snip = pkt->next;
+                if (udp_snip->type != GNRC_NETTYPE_UDP) {
+                    gnrc_pktbuf_release(pkt);
+                    break;
+                }
+
+                /* read source port and address */
+                port = byteorder_ntohs(((udp_hdr_t *)udp_snip->data)->src_port);
+
+                LL_SEARCH_SCALAR(udp_snip, ipv6_snip, type, GNRC_NETTYPE_IPV6);
+                assert(ipv6_snip != NULL);
+                src_addr = &((ipv6_hdr_t *)ipv6_snip->data)->src;
+
+                _receive(pkt, src_addr, port);
+                break;
+
+            default:
+                break;
+        }
+    }
+    return 0;
+}
+
+/* Handles incoming network IPC message. */
+static void _receive(gnrc_pktsnip_t *pkt, ipv6_addr_t *src, uint16_t port)
+{
+    coap_pkt_t coap_pkt;
+    uint8_t buf[GCOAP_RESPONSE_BUF_SIZE];
+    size_t hdr_len;
+    gnrc_pktsnip_t *resp_snip;
+
+    /* Copy request into temporary buffer, and parse it as CoAP. */
+    memcpy(buf, pkt->data, pkt->size);
+
+    int result = coap_parse(&coap_pkt, buf, pkt->size);
+    if (result < 0) {
+        DEBUG("gcoap: parse failure: %d\n", result);
+        goto exit;
+    }
+
+    if (coap_get_code_class(&coap_pkt) != COAP_CLASS_REQ) {
+        puts("gcoap: not a CoAP request.");
+        goto exit;
+    }
+
+    /* Find path for CoAP pkt among listener endpoints and execute callback. */
+    gcoap_listener_t *listener = _coap_state.listeners;
+    while (listener) {
+        coap_endpoint_t *endpoint = listener->endpoints;
+        for (size_t i = 0; i < listener->endpoints_len; i++) {
+            if (i)
+                endpoint++;
+
+            int res = strcmp((char*)&coap_pkt.url, endpoint->path);
+            if (res < 0) {
+                continue;
+            } else if (res > 0) {
+                /* endpoints expected in alphabetical order */
+                break;
+            } else {
+                if (endpoint->handler(&coap_pkt, &buf[0], pkt->size) < 0)
+                    gcoap_resp_header(&coap_pkt, &buf[0],
+                                                 GCOAP_RESPONSE_BUF_SIZE,
+                                                 COAP_CODE_INTERNAL_SERVER_ERROR);
+                goto send;
+            }
+        }
+        listener = listener->next;
+    }
+    /* endpoint not found */
+    gcoap_resp_header(&coap_pkt, &buf[0], GCOAP_RESPONSE_BUF_SIZE,
+                                          COAP_CODE_PATH_NOT_FOUND);
+
+send:
```

Comment:
### GNRC vs. sock
This send: label is where the response packet is finalized and sent to the transport layer. Here you can see how GNRC is actually used, and how we can support sock in the future.

1. _finalize_response() finalizes the buffer after the callback. 
2. Then we allocate a GNRC pktsnip and copy from the buffer into the snip. We use two memcpy() calls because the packet bytes from the callback include space between the header and the payload so _finalize_response() can write the options in order.
3. Finally _send() creates the rest of the packet buffer snips and dispatches.

To support sock, we need to write the final packet contents in the buffer rather than in the pktsnip. In this case, we can use memmove() on the buffer. Then the buffer can be passed to sock as a whole.